image-loader/test/scala/model/ImageUploadTest.scala

```diff
@@ -0,0 +1,177 @@
+      storeOrProjectOptimisedPNG
+    )
+
+    val tempFile = ResourceHelpers.fileAt(fileName)
+    val ul = UploadInfo(None)
+
+    val uploadRequest = UploadRequest(
+      uuid,
+      randomId,
+      tempFile,
+      MimeTypeDetection.guessMimeType(tempFile).right.toOption,
+      DateTime.now(),
+      "uploadedBy",
+      Map(),
+      ul
+    )
+
+    val futureImage = Uploader.uploadAndStoreImage(
+      mockDependencies.storeOrProjectOriginalFile,
+      mockDependencies.storeOrProjectThumbFile,
+      mockDependencies.storeOrProjectOptimisedImage,
+      OptimiseWithPngQuant,
+      uploadRequest,
+      mockDependencies,
+      FileMetadata()
+    )
+
+    // Assertions; Failure will auto-fail
+    futureImage.map(i => {
+      // Assertions on original request
+      assert(i.id == randomId, "Correct id comes back")
+      assert(i.source.mimeType.contains(expectedOriginalMimeType), "Should have the correct mime type")
+
+      // Assertions on generated thumbnail image
+      assert(i.thumbnail.isDefined, "Should always create a thumbnail")
+      assert(i.thumbnail.get.mimeType.get == Jpeg, "Should have correct thumb mime type")
+
+      // Assertions on optional generated optimised png image
+      assert(i.optimisedPng.isDefined == expectOptimisedFile, "Should have optimised file")
+      assert(!expectOptimisedFile || i.optimisedPng.flatMap(p => p.mimeType).contains(Png), "Should have correct optimised mime type")
+    })
+  }
+
+  ignore("rubbish") {
+    imageUpload("rubbish.jpg", Jpeg)
+  }
+  ignore("lighthouse") {
+    imageUpload("lighthouse.tif", Tiff, expectOptimisedFile = true)
+  }
+  ignore("tiff_8bpc_layered_withTransparency") {
+    imageUpload("tiff_8bpc_layered_withTransparency.tif", Tiff, expectOptimisedFile = true)
+  }
+  ignore("tiff_8bpc_flat") {
+    imageUpload("tiff_8bpc_flat.tif", Tiff, expectOptimisedFile = true)
+  }
+  ignore("IndexedColor") {
+    imageUpload("IndexedColor.png", Png)
+  }
+  ignore("bgan6a16_TrueColorWithAlpha_16bit") {
+    imageUpload("bgan6a16_TrueColorWithAlpha_16bit.png", Png, expectOptimisedFile = true)
+  }
+  ignore("basn2c16_TrueColor_16bit") {
+    imageUpload("basn2c16_TrueColor_16bit.png", Png, expectOptimisedFile = true)
+  }
+  ignore("not an image but looks like one") {
+    imageUpload("thisisnotanimage.jpg", Png, expectOptimisedFile = true).transformWith{
+      case Success(_) => fail("Should have thrown an error")
+      case Failure(e) => e match {
+        case e: ImageProcessingException => assert(e.getMessage == "File format could not be determined")
+      }
+    }
+  }
+  ignore("not an image and does not look like one") {
+    // this exception is thrown before the futures are resolved, and so does not need transformWith
+    try {
+      imageUpload("thisisnotanimage.stupid", Png, expectOptimisedFile = true)
+      fail("Should have thrown an error")
+    } catch {
+      case e: Exception => assert(e.getMessage == "No idea what you have given me")
+    }
```

Comment:
Idiomatically I'd expect this to use the scalatest matcher expect notation to test for a thrown exception but I've not done that with an AsyncXxxSuite so don't know if that works.

Reply:
Done.